### PR TITLE
feat: metadata_to_jsonstr/protocolstr and metadata_from_str

### LINKED ISSUE #82: at_c: atclient_atkey_metadata_from_string
Populate the *metadata struct given the corresponding arguments.

Located in `atclient/src/metadata.c`

```c
int atclient_atkey_metadata_from_string(atclient_atkey_metadata *metadata, const char *metadatastr, const unsigned long metadatastrlen)
{
    return 1; // TODO: implement
}
```

### LINKED ISSUE #83: at_c: atclient_atkey_metadata_to_string
Take an already populated *metadata struct and write it to a buffer string *metadatastr

`atclient/src/metadata.c`

```c
int atclient_atkey_metadata_to_string(atclient_atkey_metadata *metadata, char *metadatastr, const unsigned long metadatastrlen, unsigned long *metadatastrolen)
{
    return 1; // TODO: implement
}
```

### PR DESCRIPTION
**Pull request by JeremyTubongbanua**
closes #82 
closes #83

**- What I did**
- Implemented all of metadata.c, namely these functions:
- `int atclient_atkey_metadata_from_jsonstr(atclient_atkey_metadata *metadata, const char *metadatastr, const size_t metadatastrlen);`
- `int atclient_atkey_metadata_to_jsonstr(const atclient_atkey_metadata metadata, char *metadatastr, const size_t metadatastrlen, size_t *metadatastrolen);`
- `int atclient_atkey_metadata_to_protocolstr(const atclient_atkey_metadata metadata, char *metadatastr, const size_t metadatastrlen, size_t *metadatastrolen);`

**- How to verify it**

New unit tests:
- `test_atkey_metadata.c`

**- Description for the changelog**
Other misc changes
- refactored booleans (from integers to std bools) in metadata.c/.h
- ctest now only outputs on failure (better printing cause the amount of tests we have is growing)
- removed useless null check on an integer
- fixed atstr major bug
- atstr now initializes to zeroes

### CODE REVIEW

File: packages/atclient/include/atclient/metadata.h [hunk 1]
```diff
@@ -3,59 +3,172 @@
...
+#define DATE_STR_BUFFER_SIZE 256 // can hold most date strings found in metadata
+#define GENERAL_BUFFER_SIZE 8192 // can hold most metadata strings
 
 typedef struct atclient_atkey_metadata {
-  atclient_atstr createdby; // read and set by the protocol
-  atclient_atstr updatedby; // read and set by the protocol
-  atclient_atstr createdat; // date representing when key was created
-  atclient_atstr updatedat; // date representing when key was last modified
-  atclient_atstr status;
+  atclient_atstr createdby; // represents the atSign who created this atkey. read from protoocl string
```
Comment (xavierchanth) [inline packages/atclient/include/atclient/metadata.h]: `protocol` typo

File: packages/atclient/include/atclient/metadata.h [hunk 2]
```diff
@@ -3,59 +3,172 @@
...
+  atclient_atstr updatedby; // repreesnts the atSign who last updated this atkey. read from protocol string only
```
Comment (xavierchanth) [inline packages/atclient/include/atclient/metadata.h]: `represents` typo

Comment (JeremyTubongbanua) [reply packages/atclient/include/atclient/metadata.h]: Completely redid the documentation https://github.com/atsign-foundation/at_c/pull/117/commits/4e8571eceb5e5153954040a451df42337e82a4c6

File: packages/atclient/include/atclient/metadata.h (hunk 1 above)
Comment (JeremyTubongbanua) [reply packages/atclient/include/atclient/metadata.h]: Completely redid the documentation https://github.com/atsign-foundation/at_c/pull/117/commits/4e8571eceb5e5153954040a451df42337e82a4c6

File: packages/atclient/include/atclient/metadata.h [hunk 3]
```diff
@@ -2,60 +2,363 @@
...
-#define GENERAL_BUFFER_LENGTH 8192 // can hold most things found in metadata
+// initializedfields[0]
+#define ATKEY_METADATA_CREATEDBY_INITIALIZED 0b00000001
+#define ATKEY_METADATA_UPDATEDBY_INITIALIZED 0b00000010
+#define ATKEY_METADATA_STATUS_INITIALIZED 0b00000100
+#define ATKEY_METADATA_VERSION_INITIALIZED 0b00001000
+#define ATKEY_METADATA_EXPIRESAT_INITIALIZED 0b00010000
+#define ATKEY_METADATA_AVAILABLEAT_INITIALIZED 0b00100000
+#define ATKEY_METADATA_REFRESHAT_INITIALIZED 0b01000000
+#define ATKEY_METADATA_CREATEDAT_INITIALIZED 0b10000000
```
Comment (xavierchanth) [inline packages/atclient/include/atclient/metadata.h:19]: would be cleaner to use bit shift syntax next time (1 << 0), (1 << 1), etc.. to help avoid small typos

File: packages/atclient/src/atkey.c [hunk 4]
```diff
@@ -88,7 +89,7 @@ int atclient_atkey_from_string(atclient_atkey *atkey, const char *atkeystr, cons
   }
   // check if cached
   if (strncmp(token, "cached", strlen("cached")) == 0) {
-    atkey->metadata.iscached = 1;
+    atkey->metadata.iscached = true;
```
Comment (xavierchanth) [inline packages/atclient/src/atkey.c:92]: do we need to make sure the initialized vector is set here?

**Review by xavierchanth — APPROVED** [review]

**Pull request merged**